### ISSUE TITLE
playbook incorrectly targets localhost for all plays after calls to local_action and add_host

### ISSUE THREAD

**Issue by mbking1269** [ansible/ansible#16898]
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

I have made no alterations to the ansible configuration.
##### OS / ENVIRONMENT

Ansible is running on Ubuntu 14.04. The target host is CentOS 6.7.
##### SUMMARY

After a combination of local_action and add_host, plays with hosts: all incorrectly run tasks on localhost after the add_host call. Localhost was never specified in my inventory hosts file. If I remove either of these tasks (local_action or add_host), plays/tasks correctly run only on the hosts specified in the inventory.
##### STEPS TO REPRODUCE

I have created a simple playbook to illustrate this issue. I have pasted the contents of the playbook and the hosts file below. The hosts file has just one target, deployTarget. I expect all tasks except local_action to run only on deployTarget.

```
Playbook:

---
- name: "Host check 1"
  hosts: all
  become: false
  tasks:
   - name: Echo Hello 
     local_action: shell echo hello
     changed_when: no

- name: "Gather facts 1"
  hosts: all
  gather_facts: yes
  become: no

- name: "Host check 2"
  hosts: all
  become: false
  roles:

- name: "Build host group and do something with it"
  hosts: host_group_two
  become: no
  gather_facts: no

  tasks:
    - name: Build up collection of hosts
      add_host:
        name: "{{ item }}"
        group: new_host_group 
      with_items:
        - "{{ groups['host_group_one'] }}"
        - "{{ groups['host_group_two'] }}"
      changed_when: no

- name: "Gather facts 2"
  hosts: all
  gather_facts: yes
  become: no

- name: "Host check 3"
  hosts: all
  become: false
  roles:


hosts:
deployTarget ansible_ssh_host=x.x.x.x ansible_ssh_user=<user> ansible_ssh_private_key_file=<key>

[host_group_one]
deployTarget

[host_group_two]
deployTarget
```
##### EXPECTED RESULTS

Below is the expected output showing the playbooking performing tasks just on deployTarget with the only exception being the local_action task, 'Echo Hello'.
PLAY [Host check 1] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

TASK [Echo Hello] **************************************************************
ok: [deployTarget -> localhost]

PLAY [Gather facts 1] **********************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY [Host check 2] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY [Build host group and do something with it] *******************************

TASK [Build up collection of hosts] ********************************************
ok: [deployTarget] => (item=deployTarget)
ok: [deployTarget] => (item=deployTarget)

PLAY [Gather facts 2] **********************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY [Host check 3] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY RECAP *********************************************************************
deployTarget               : ok=7    changed=0    unreachable=0    failed=0    
##### ACTUAL RESULTS

Note the incorrect injection of localhost after the TASK 'Build up collection of hosts'. Given the hosts file does not call out localhost nor does the add_host task explicitly include localhost, I would not expect any tasks to run on localhost.

PLAY [Host check 1] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

TASK [Echo Hello] **************************************************************
ok: [deployTarget -> localhost]

PLAY [Gather facts 1] **********************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY [Host check 2] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]

PLAY [Build host group and do something with it] *******************************

TASK [Build up collection of hosts] ********************************************
ok: [deployTarget] => (item=deployTarget)
ok: [deployTarget] => (item=deployTarget)

PLAY [Gather facts 2] **********************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]
ok: [localhost]

PLAY [Host check 3] ************************************************************

TASK [setup] *******************************************************************
ok: [deployTarget]
ok: [localhost]

PLAY RECAP *********************************************************************
deployTarget               : ok=7    changed=0    unreachable=0    failed=0  
localhost  

```
See results above for command output.
```

**Comment by bcoca** [comment]
Confirmed, looking at fix to remove implicit localhost from group 'all'.

**Comment by bcoca** [comment]
fixed via #16900
